### PR TITLE
Fix distance snapping grid providing incorrect time values for non-1.0x spacing

### LINKED ISSUE #18073: Distance snapping looks to be broken
### Type

Game behaviour

### Bug description

When set to non-1.0x, the distance snap grid provided snaps do not align correctly:

osu! 2022-05-03 at 09 26 55

As an aside, there's also no way to reset the distance snap to default:

osu! 2022-05-03 at 09 26 09

### Screenshots or videos

_No response_

### Version

master

### Logs

-

### PR DESCRIPTION
**Pull request by peppy**
- Alternative to https://github.com/ppy/osu/pull/18092
- Closes https://github.com/ppy/osu/issues/18073

Of note, only 4c884aea5d is required to fix the issue. The majority of the work here was to rewrite the snapping implementation in `CircularDistanceSnapGrid` to use the snap provider, rather than calculate things locally using its own implementation (which felt very ad-hoc).

- [x] Depends on https://github.com/ppy/osu/pull/18100
- [x] Depends on https://github.com/ppy/osu/pull/18101

### CODE REVIEW

File: osu.Game/Screens/Edit/Compose/Components/DistanceSnapGrid.cs [hunk 1]
```diff
@@ -23,7 +23,9 @@ public abstract class DistanceSnapGrid : CompositeDrawable
         /// <summary>
         /// The spacing between each tick of the beat snapping grid.
         /// </summary>
-        protected float DistanceSpacing { get; private set; }
+        protected float DistanceBetweenTick { get; private set; }
```
Comment (bdach) [inline osu.Game/Screens/Edit/Compose/Components/DistanceSnapGrid.cs:26]: Probably should be renamed to `DistanceBetweenTicks` (as in, plural ticks). "between tick" sounds a touch weird.

**Comment by bdach** [comment]
There's a weird issue with the new snapping logic:

https://user-images.githubusercontent.com/20418176/167080863-a9e63637-9cfb-4d7f-8d97-de112326ed8d.mp4

* The sudden jump in distance looks wrong.
* After I start dragging far away, the distance looks to be clamped, but the time values are not.

**Comment by peppy** [comment]
Nice find. Part of the issue was that the `StartTime` of the reference object was being used instead of `GetEndTime`, and the second part is that the snapped duration wasn't recomputed when clamping the distance to the extent.

The tests don't really lend to testing these easily so I haven't updated the tests.

**Comment by peppy** [comment]
I've fixed the remaining issues pointed out. Hopefully the changes make sense mathematically? We probably need a completely new test scene to cover this (one which covers the snap values) but I don't know how much value there will be in that, considering how rarely this code will change. Maybe if it regresses post merge I'll look into that?

**Comment by bdach** [comment]
Appears to work well in my testing now, as far as I can tell :+1:

**Review by bdach — DISMISSED** [review]

**Comment by frenzibyte** [comment]
This generally looks and reads pretty well. I've played around with the grid and noticed this issue when the hitobject time is not snapped (due to changing beat divisor):

https://user-images.githubusercontent.com/22781491/167146905-83259653-8ded-401b-a8b3-6d673b1ce4e0.mp4

Not entirely sure whether it is the grid rings or the position snap logic that is at fault here (I'm also not sure whether hitobjects with unsnapped times are even valid).

**Comment by bdach** [comment]
Isn't distance snap grid already half broken on master with unsnapped objects? Not sure how relevant that is in the context of this pull.

Not to mention that to me it looks that the grid is drawn wrong but the snapped positions are correct.

**Comment by frenzibyte** [comment]
Upon a [quick explaination](https://discord.com/channels/188630481301012481/188630652340404224/972138461920378920) regarding what's going on in that video and the fact that master already has time-unsnapped hitobjects broken, it's fine to dismiss the above issue and resolve it separately if it matters.

**Review by frenzibyte — APPROVED** [review]

**Pull request merged**